### PR TITLE
Limit displayed Water Quality catchments to those intersecting & within the shape

### PR DESCRIPTION
**Pull request by kellyi**
## Overview

This PR adjusts the `catchment_water_quality`'s PostGIS query to return only catchments which intersect by some (small) percentage of their total area: 1% for AOI's less than about 1000 sq km; 3% for AOIs > than that.

The outcome's that we now don't display some catchments which intersect but only abut the AOI. This was most visible for the RWD-generated shapes, but it also impacted HUCs.

Connects #2240 

### Demo

rwd-wq-catchments

### Notes

We could probably set the percent limit to just be 3 or 5% for all aois and have this work the same way, but I wanted to make sure we didn't unnecessarily exclude some catchments which should be displayed.

Not sure how the new PostGIS query performs compared with the prior `ST_Intersects`.

## Testing Instructions
- get this branch, set up RWD, then bundle
- use Continental US RWD to select points within the DRB and verify that the shapes now don't include any non-intersecting shapes -- and that the totality of the shapes which are returned encompasses the AOI

### CODE REVIEW

**Comment by kellyi** [comment]
One other thing I noticed when working on this is that it looks like we've got duplicate `calcs.py` files: one in the geoprocessing dir, another in modeling: https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/apps/modeling/calcs.py

The model's version of the water quality function, at least, didn't seem to get called. If the whole file's no longer relevant we might make plans to remove it altogether.

**Comment by arottersman** [comment]
- Water quality is noticeably slower, but I suppose that's to be expected

- Really small shapes don't find intersections:
screen shot 2017-10-19 at 1 11 22 pm

screen shot 2017-10-19 at 1 12 03 pm

**Comment by mmcfarland** [comment]
> Water quality is noticeably slower, but I suppose that's to be expected

I noticed this too, but would like to quantify it by timing the query.  It's possible, depending on the plan, that post-processing the results with GEOS could be faster.  But I'd like to see how much longer the adjusted query takes before we try out alternate implementation.

**Comment by kellyi** [comment]
Just so I don't forget: as in #2343, we should reproject the shape with GEOS to 5070 so we can use meter values.

**Comment by kellyi** [comment]
I decided to revert the PostGIS statement changes and handle this on the app side using GEOSGeometry, since it'll be simpler to work out the logic there. I didn't precisely track the performance difference, but my impression's that even with this additional check the WQ catchments still take far less time than the geoprocessing ops.

16c7ac8 sets up the implementation, and it currently selects only catchments for which the area of intersection with the AOI represents > 1% of the catchment's area or > 1% of the AOI's area. This works for *most* cases but I'm going to try tweaking the values and using `and` instead of `or` to see if that gives better results.

**Comment by kellyi** [comment]
This should be ready for another look. After testing out some minimum intersection percentage values, I decided to use a sliding scale which clamps it between 1 and 5% based on the size of the AOI. I did this because huge AOIs often dwarf the catchments in size, so visually small intersection areas ended up being relatively large percentages of the catchment.

For AOIs < 1000 sq km, the minimum intersection area is 1% of the catchment's total area. For AOIs > 1000 sq km, the value is the AOI's sq km value / 1000 up to a maximum of 5%. If we keep this implementation, we may want to dial that down further -- to 3% or maybe even less.

File: src/mmw/apps/geoprocessing_api/calcs.py [hunk 1]
```diff
@@ -135,8 +135,33 @@ def catchment_water_quality(geojson):
             ]
         else:
             catchment_water_quality_results = []
+
+    def catchment_intersects_aoi(aoi, geom):
```
Comment (mmcfarland) [inline src/mmw/apps/geoprocessing_api/calcs.py]: I'd recommend taking this function out of the containing function, as it's a little hard to follow the flow of the outer function.  The var `geom` is declared in both the inner and outer function, and in the outer scope references to the aoi, but in the inner scope references an individual catchment result.  I think, given that there are multiple geometries being handled here, we should abandon `geom` as a var name since it's not clear which it refers to.

File: src/mmw/apps/geoprocessing_api/calcs.py [hunk 2]
```diff
@@ -135,8 +135,33 @@ def catchment_water_quality(geojson):
...
+        # Clamp minimum intersection percentage for the catchment between 1 and
+        # 5%, with the percentage sliding from 1 up to 5 if the AOI's > 1000
+        # sq km in size (deriving from the AOI sq km value / 1000).
+        aoi_kms = aoi.area / 1000000
+        min_intersection_pct = 1 if aoi_kms < 1000 else min(5, aoi_kms / 1000)
+        catchment_geom = GEOSGeometry(json.dumps(geom), srid=4326)
+        catchment_geom.transform(5070, clone=True)
```
Comment (mmcfarland) [inline src/mmw/apps/geoprocessing_api/calcs.py]: Since you're setting the `clone` arg, you'll need to capture the return variable.  I don't believe it mutates the object you're calling `transform` on.

File: src/mmw/apps/geoprocessing_api/calcs.py [hunk 3]
```diff
@@ -135,8 +135,33 @@ def catchment_water_quality(geojson):
...
+
+        if not catchment_geom.valid:
+            return False
+
+        if not catchment_geom.intersects(aoi):
```
Comment (mmcfarland) [inline src/mmw/apps/geoprocessing_api/calcs.py]: Every DRB AoI I tested wouldn't intersect with any results, I believe because `catchment_geom` is not `transformed` in place.

File: src/mmw/apps/geoprocessing_api/calcs.py [hunk 4]
```diff
@@ -135,8 +135,33 @@ def catchment_water_quality(geojson):
...
+            return False
+
+        intersection_pct = (catchment_geom.intersection(aoi).area /
+                            catchment_geom.intersection(aoi).area * 100)
+
+        return intersection_pct > min_intersection_pct
+
+    geom.transform(5070)
```
Comment (mmcfarland) [inline src/mmw/apps/geoprocessing_api/calcs.py]: Consider setting `clone` on this and referencing the result in a var that distinguishes it from the input.  Also consider transforming it immediately after creating it, so it's clear what its contents are.

File: src/mmw/apps/geoprocessing_api/calcs.py [hunk 5]
```diff
@@ -135,8 +135,33 @@ def catchment_water_quality(geojson):
...
+        catchment_geom.transform(5070, clone=True)
+
+        if not catchment_geom.valid:
+            return False
+
+        if not catchment_geom.intersects(aoi):
+            return False
+
+        intersection_pct = (catchment_geom.intersection(aoi).area /
+                            catchment_geom.intersection(aoi).area * 100)
```
Comment (kellyi) [inline src/mmw/apps/geoprocessing_api/calcs.py]: Just realized the math is wrong here -- should be `/ catchment_geom.area * 100`.

File: src/mmw/apps/geoprocessing_api/calcs.py (hunk 1 above)
Comment (kellyi) [reply src/mmw/apps/geoprocessing_api/calcs.py]: Sounds good -- I moved this out into a separate function with better argument names, and also passed in a cloned version of the area of interest geometry in 5070.

**Comment by kellyi** [comment]
> I would suggest popping the function up a level and adding a few unit tests showing how it behaves under the various conditions you've included in your clamp function.

I moved the function out to the top level in https://github.com/WikiWatershed/model-my-watershed/pull/2383/commits/a09ed54309de070f06198626efec47cbb81cb4f8 and also adjusted the names and the GEOS transformations.

That function keeps the clamped 1-5% min intersection value, and includes the catchment if the intersection area represents an % of the catchment or aoi total area that's greater than the min and the total of those percentages is > 2%. This was done because for some aois there were catchments for which the % was 1.05 or something of the catchment but 0.01% or so of the AOI.

This value should probably be tweaked and it may be worthwhile to derive it from the sliding min intersection % somehow. Likewise it may be worthwhile to limit the top of the min intersection % scale to 3% rather than 5%.

**Comment by kellyi** [comment]
I rebased this on `develop` and squashed it down to a few commits: the first one's left here just temporarily for documentary purposes; the second one reverts it and performs the check via a `check_catchment_intersects_aoi` function. Third commit is a set of unit tests for this which encompass checking abutting, intersecting, and contained catchments the following sets of test data.

For each test:

- the *abutting_catchment* should be `False`, since it doesn't properly intersect
- the *intersecting_catchment* should be `True`, since it does
- the *contained* or *containing* catchment should be `True`

## AOI of about 1 sq km

screen shot 2017-11-01 at 5 37 46 pm

<details>
<summary>
AOI about 1 sq km
</summary>

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -75.27900695800781,
              39.891925022904516
            ],
            [
              -75.26608943939209,
              39.891925022904516
            ],
            [
              -75.26608943939209,
              39.90173657727282
            ],
            [
              -75.27900695800781,
              39.90173657727282
            ],
            [
              -75.27900695800781,
              39.891925022904516
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -75.27368545532225,
              39.89722607068418
            ],
            [
              -75.26887893676758,
              39.89722607068418
            ],
            [
              -75.26887893676758,
              39.90124274066003
            ],
            [
              -75.27368545532225,
              39.90124274066003
            ],
            [
              -75.27368545532225,
              39.89722607068418
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -75.28535842895508,
              39.898279646242635
            ],
            [
              -75.27896404266357,
              39.898279646242635
            ],
            [
              -75.27896404266357,
              39.90305345750681
            ],
            [
              -75.28535842895508,
              39.90305345750681
            ],
            [
              -75.28535842895508,
              39.898279646242635
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -75.26849269866943,
              39.890838422106924
            ],
            [
              -75.26244163513184,
              39.890838422106924
            ],
            [
              -75.26244163513184,
              39.89498716884207
            ],
            [
              -75.26849269866943,
              39.89498716884207
            ],
            [
              -75.26849269866943,
              39.890838422106924
            ]
          ]
        ]
      }
    }
  ]
}
```

</details>

## AOI of about 100 sq km

screen shot 2017-11-01 at 5 39 11 pm

<details>
<summary>
AOI About 100 sq km
</summary>

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.64584350585938,
              38.96154447940714
            ],
            [
              -94.53460693359374,
              38.96154447940714
            ],
            [
              -94.53460693359374,
              39.05225165582583
            ],
            [
              -94.64584350585938,
              39.05225165582583
            ],
            [
              -94.64584350585938,
              38.96154447940714
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.53563690185547,
              39.03065255999985
            ],
            [
              -94.49203491210938,
              39.03065255999985
            ],
            [
              -94.49203491210938,
              39.07864158248181
            ],
            [
              -94.53563690185547,
              39.07864158248181
            ],
            [
              -94.53563690185547,
              39.03065255999985
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.55554962158203,
              38.92870117926206
            ],
            [
              -94.49581146240233,
              38.92870117926206
            ],
            [
              -94.49581146240233,
              38.9858333874019
            ],
            [
              -94.55554962158203,
              38.9858333874019
            ],
            [
              -94.55554962158203,
              38.92870117926206
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.62284088134766,
              38.997841307500714
            ],
            [
              -94.58576202392578,
              38.997841307500714
            ],
            [
              -94.58576202392578,
              39.031452644263084
            ],
            [
              -94.62284088134766,
              39.031452644263084
            ],
            [
              -94.62284088134766,
              38.997841307500714
            ]
          ]
        ]
      }
    }
  ]
}
```

</details>

## AOI of about 1000 sq km

screen shot 2017-11-01 at 5 40 23 pm

<details>
<summary>
AOI about 1000 sq km
</summary>

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -96.1083984375,
              41.12074559016745
            ],
            [
              -95.7513427734375,
              41.12074559016745
            ],
            [
              -95.7513427734375,
              41.39741506646461
            ],
            [
              -96.1083984375,
              41.39741506646461
            ],
            [
              -96.1083984375,
              41.12074559016745
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -95.8172607421875,
              41.0607151401866
            ],
            [
              -95.68405151367188,
              41.0607151401866
            ],
            [
              -95.68405151367188,
              41.160046141686905
            ],
            [
              -95.8172607421875,
              41.160046141686905
            ],
            [
              -95.8172607421875,
              41.0607151401866
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -95.93811035156249,
              41.306697618181865
            ],
            [
              -95.82550048828125,
              41.306697618181865
            ],
            [
              -95.82550048828125,
              41.3757780692323
            ],
            [
              -95.93811035156249,
              41.3757780692323
            ],
            [
              -95.93811035156249,
              41.306697618181865
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -96.18255615234375,
              41.24064190269475
            ],
            [
              -96.10736846923828,
              41.24064190269475
            ],
            [
              -96.10736846923828,
              41.2765163855178
            ],
            [
              -96.18255615234375,
              41.2765163855178
            ],
            [
              -96.18255615234375,
              41.24064190269475
            ]
          ]
        ]
      }
    }
  ]
}
```

</details>

## AOI of about 10000 sq km

screen shot 2017-11-01 at 5 41 10 pm

<details>
<summary>
AOI about 10k sq km
</summary>

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -115.01586914062499,
              43.866218006556394
            ],
            [
              -113.719482421875,
              43.866218006556394
            ],
            [
              -113.719482421875,
              44.89479576469787
            ],
            [
              -115.01586914062499,
              44.89479576469787
            ],
            [
              -115.01586914062499,
              43.866218006556394
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -115.23559570312499,
              44.380802793578475
            ],
            [
              -115.00488281250001,
              44.380802793578475
            ],
            [
              -115.00488281250001,
              44.52001001133986
            ],
            [
              -115.23559570312499,
              44.52001001133986
            ],
            [
              -115.23559570312499,
              44.380802793578475
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -115.17791748046875,
              43.775060351224695
            ],
            [
              -114.949951171875,
              43.775060351224695
            ],
            [
              -114.949951171875,
              44.09350315285847
            ],
            [
              -115.17791748046875,
              44.09350315285847
            ],
            [
              -115.17791748046875,
              43.775060351224695
            ]
          ]
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -114.43359375,
              44.262904233655384
            ],
            [
              -114.06829833984375,
              44.262904233655384
            ],
            [
              -114.06829833984375,
              44.61393394730626
            ],
            [
              -114.43359375,
              44.61393394730626
            ],
            [
              -114.43359375,
              44.262904233655384
            ]
          ]
        ]
      }
    }
  ]
}
```

</details>

## A very large AOI with very small catchments

screen shot 2017-11-01 at 5 43 45 pm

<details>
<summary>
large AOI, small catchments
</summary>

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
            "type": "Polygon",
            "coordinates": [
                [
                    [
                        -85.166015625,
                        39.470125122358176
                    ],
                    [
                        -82.44140625,
                        39.470125122358176
                    ],
                    [
                        -82.44140625,
                        42.94033923363181
                    ],
                    [
                        -85.166015625,
                        42.94033923363181
                    ],
                    [
                        -85.166015625,
                        39.470125122358176
                    ]
                ]
            ]
        }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
            "type": "Polygon",
            "coordinates": [
                [
                    [
                        -85.440673828125,
                        42.68243539838623
                    ],
                    [
                        -85.15502929687499,
                        42.68243539838623
                    ],
                    [
                        -85.15502929687499,
                        42.79540065303723
                    ],
                    [
                        -85.440673828125,
                        42.79540065303723
                    ],
                    [
                        -85.440673828125,
                        42.68243539838623
                    ]
                ]
            ]
        }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
            "type": "Polygon",
            "coordinates": [
                [
                    [
                        -82.63916015625,
                        41.94314874732696
                    ],
                    [
                        -82.265625,
                        41.94314874732696
                    ],
                    [
                        -82.265625,
                        42.06560675405716
                    ],
                    [
                        -82.63916015625,
                        42.06560675405716
                    ],
                    [
                        -82.63916015625,
                        41.94314874732696
                    ]
                ]
            ]
        }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
            "type": "Polygon",
            "coordinates": [
                [
                    [
                        -83.671875,
                        39.65645604812829
                    ],
                    [
                        -83.34228515625,
                        39.65645604812829
                    ],
                    [
                        -83.34228515625,
                        39.9434364619742
                    ],
                    [
                        -83.671875,
                        39.9434364619742
                    ],
                    [
                        -83.671875,
                        39.65645604812829
                    ]
                ]
            ]
        }
    }
  ]
}
```
</details>

## A very small AOI with very large catchments

screen shot 2017-11-01 at 5 45 42 pm

<details>
<summary>
small AOI, large catchments
</summary>

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
            "type": "Polygon",
            "coordinates": [
                [
                    [
                        -86.1189079284668,
                        30.712618489700507
                    ],
                    [
                        -86.11066818237303,
                        30.712618489700507
                    ],
                    [
                        -86.11066818237303,
                        30.719554693895116
                    ],
                    [
                        -86.1189079284668,
                        30.719554693895116
                    ],
                    [
                        -86.1189079284668,
                        30.712618489700507
                    ]
                ]
            ]
        }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
            "type": "Polygon",
            "coordinates": [
                [
                    [
                        -86.11856460571288,
                        30.71940712027702
                    ],
                    [
                        -86.12113952636719,
                        30.88395860861961
                    ],
                    [
                        -86.38206481933594,
                        30.884547891921986
                    ],
                    [
                        -86.37931823730467,
                        30.71586528568626
                    ],
                    [
                        -86.11856460571288,
                        30.71940712027702
                    ]
                ]
            ]
        }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
            "type": "Polygon",
            "coordinates": [
                [
                    [
                        -86.13006591796874,
                        30.59832078510471
                    ],
                    [
                        -85.9075927734375,
                        30.59832078510471
                    ],
                    [
                        -85.9075927734375,
                        30.714094319607913
                    ],
                    [
                        -86.13006591796874,
                        30.714094319607913
                    ],
                    [
                        -86.13006591796874,
                        30.59832078510471
                    ]
                ]
            ]
        }
    },
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
            "type": "Polygon",
            "coordinates": [
                [
                    [
                        -86.22550964355469,
                        30.627277165616874
                    ],
                    [
                        -86.0394287109375,
                        30.627277165616874
                    ],
                    [
                        -86.0394287109375,
                        30.80967992229391
                    ],
                    [
                        -86.22550964355469,
                        30.80967992229391
                    ],
                    [
                        -86.22550964355469,
                        30.627277165616874
                    ]
                ]
            ]
        }
    }
  ]
}
```

</details>

**Comment by mmcfarland** [comment]
A reminder to retarget the release branch before merging.

**Comment by kellyi** [comment]
Thanks for all your help with this! I'll rebase and retarget etc before merging.
